Integration 4.0.8

dostuffthatmatters opened the pull request:
**Integration still in progress:**

- [x] Freeze the poetry dependencies to the one used in `4.0.7`
- [x] Remove git commit SHA from emails
- [x] Closes #151
- [x] Closes #126 
- [x] Closes #125 
- [x] Closes #154
- [x] Closes #157 
- [x] Closes #165
- [x] Closes #156 
- [x] Closes #167
- [x] Close #169

dostuffthatmatters commented:
Had to rebase it onto the main branch, that is why I force-pushed.

patrickjaigner commented on packages/cli/commands/core.py:
```diff
@@ -61,14 +71,22 @@ def _start_pyra_core() -> None:
+    try:
+        _run_pyra_core_lock.acquire()
+        _run_pyra_core_lock.release()
+    except filelock.Timeout:
```
Is this the best way to do this?

dostuffthatmatters replied:
I think so.

`run_pyra_core.py` will acquire that lock before the mainloop starts and frees it when the program finishes or crashes: https://github.com/tum-esm/pyra/blob/52eb662f0bfb33d224daf9cf7cfe6cce192da732/run_pyra_core.py#L14-L22

When the `pyra-cli start` command cannot acquire the Lock, some Pyra Core process must be running.

The other option would be to cycle through all running processes (like the `is-running` command), but I think that is more work to implement because there are different cases.

patrickjaigner commented on packages/core/modules/enclosure_control.py:
```diff
@@ -235,19 +236,18 @@ def auto_set_power_spectrometer(self) -> None:
-            self.config["general"]["min_sun_elevation"] - 1
-        ) * utils.Astronomy.units.deg
-
-        if current_sun_elevation is not None:
```
Why could you remove this line?

dostuffthatmatters replied:
There is also a new test that draw random datetimes from each year between 2020 and 2050 and checks if the Astronomy class can generate an elevation for that: https://github.com/tum-esm/pyra/blob/integration-4.0.8/tests/utils/test_astronomy.py

patrickjaigner commented on packages/core/modules/opus_measurement.py:
Could you let me know why you changed the Docstrings? Now they are different lengths and also every shorter than the rest of the code.

dostuffthatmatters replied:
Sure! The wide comment blocks are hard to read on small monitors. I often work with two files open next to each other:

<img width="1352" alt="Screenshot 2023-05-08 at 20 42 06" src="https://user-images.githubusercontent.com/29046316/236919617-3887034c-ebf9-4691-8273-70fa4148231d.png">

But when the lines of the doc strings are very long, it is super annoying when every lines wraps to a new line:

<img width="548" alt="Screenshot 2023-05-08 at 20 41 48" src="https://user-images.githubusercontent.com/29046316/236919880-d6458122-f7e7-4b6b-a894-e8f1f77c6a3c.png">

On big monitors, that is not an issue but whenever using two files or doing any diffing stuff on a tiny one, it bothers a lot.

Would you like to revert it?

patrickjaigner replied:
Didn't realize this, but sure let's keep the changes.

patrickjaigner commented on packages/core/modules/opus_measurement.py:
```diff
@@ -17,35 +17,59 @@
+        #             logger.debug(f'Successfully closed OPUS instance "{p}"')
+        #     except Exception as e:
+        #         logger.warning(f"Failed to close OPUS: {e}")
+
```
I think we should not keep in in the code, but move it to the issue.

dostuffthatmatters commented:
> I finished reading through the changes. There is a lot of good refactoring done. I also added some comments for lines where I needed clarification on why changes were made.

Thank you for the review 👍

> Overall we should think about how easy and accessible the tum-esm-utils library is. Will people easily find the API description (a quick Google search didn't give me the correct result) or will it confuse people as relevant code is now moved to a different repository? I'm unsure if I like it for Pyra, as it will make it harder for people to add additional things to functions or make changes. What do you think?

Good question! The index of DuckDuckGo is great:

<img width="1352" alt="Screenshot 2023-05-08 at 21 52 39" src="https://user-images.githubusercontent.com/29046316/236920943-affa94e1-c1cc-483a-8674-5926b44c15dd.png">

But the index of Google is terrible 😅:

<img width="1352" alt="Screenshot 2023-05-08 at 21 54 01" src="https://user-images.githubusercontent.com/29046316/236921046-26acdc3f-9c3a-464d-8a42-44b5eb629bdb.png">

*(Both searches were done on a browser with no local traces of me)*

<br/>

**You are right about making changes harder by outsourcing thinks. Three things I would do:**

* We can add a link to the utils repository to every project's README that uses it, this way, the page rank will improve and Google should have it soon too.

* If we adhere to only outsource the things that are very general (`run_shell_command`, `merge_dicts`) and strictly type and document all functions, I think this is an improvement on the complexity of all projects that use the library. The current set of functions is already well documented I think, and most custom utility folder has worse testing and documentation than this library:

<img width="998" alt="Screenshot 2023-05-08 at 22 00 43" src="https://user-images.githubusercontent.com/29046316/236922602-f9a1f391-0d28-4a83-a298-c11042bee366.png">

* We can add a link to the repository, the website and the PyPI page to the doc string of the module, then it is quicker to find documentation about it:

<img width="998" alt="Screenshot 2023-05-08 at 22 00 14" src="https://user-images.githubusercontent.com/29046316/236923035-de6d0c45-35cd-479f-8998-c65da0a4434c.png">

dostuffthatmatters commented:
EDIT: Regarding the documentation of the utility library, from `1.5.2` (will be release in a few minutes), the module doc strings will include more information:


https://user-images.githubusercontent.com/29046316/236928577-64974ed6-f759-403e-84b4-adaa740f4ad7.mp4

patrickjaigner commented:
@dostuffthatmatters  Finished my review. All points clarified, thank you.

dostuffthatmatters commented:
@patrickjaigner this is running well on the integration station. Do you approve the release?

patrickjaigner commented:
@dostuffthatmatters Sounds good. Let's confirm a 7 days stable run of the release. Other than that I agree that 4.0.8 is ready to go.
